### PR TITLE
 new -stop-after-parse option: stop after the parsing phase 

### PR DESCRIPTION
**Pull request by gasche**
This commit was part of the gigantic Menhir-parser PR (I need it for benchmarking purpose), but:
- it is of independent interest
- I got tired or rebasing it against the surprisingly-many changes to `driver/`
- I expect some discussion on the option naming, so early debate is good

It sits on top of #1944 and serves to demonstrate why #1944 is required.

### CODE REVIEW

**Comment by ghost** [comment]
Regarding the name of the option: what about `-stop-after parse`? This way, if in the future we want to add the possibility to stop after another phase, we don't need to add a new option.

**Comment by gasche** [comment]
Yes; we had plans with @bobot to standardize pass options in this way (in particular, `-i` should be split into `-dtypes -stop-after typing`), but we never had the time to take a coherent look at it. I'm fine with `-stop-after ARG` as long as we accept to only support `parse` (or `parsing`, etc.) for now to, preferably, not overload the PR with unrelated concerns.

**Comment by ghost** [comment]
Agreed, I'm only suggesting a UI change

**Comment by chambart** [comment]
Would you consider producing an output ? It could be the marshalled ast similar to the output of camlp4.
Otherwise, this should be stated to be some kind of debug option.

**Comment by gasche** [comment]
@chambart: the way I think of it, most passes of the compiler can produce some output (parsing can produce a serialized AST, typing can infer a `.mli` file, lambda or cmm outputs could be usefully serialized), and we should have separate flags to say "please make this specific pass produce a result" and "stop after this pass". `-i` says both things at once, at this is wrong, because it is useful to do one without the other (stop after typing to get warnings, produce `.inferred.mli` and keep compiling).

The present option says to stop after parsing, and we could have a different option, for example `-dast`, to produce the serialized AST as an output.

Note that the current flag is already useful (without `-dast`) for users who want to check that their file is syntactically correct, but nothing else.

**Comment by gasche** [comment]
I just pushed a new commit that implements @diml's suggestion. I could not resist the temptation to generalize slightly, and the `-stop-after` option accepts either `parsing` or `typing` -- in the latter case, it behaves as `-i`, but without printing the interface.

File: utils/clflags.ml
Comment (ghost) [inline utils/clflags.ml]: This comment is no longer true

**Review by ghost — CHANGES_REQUESTED** [review]

File: driver/main_args.ml
Comment (xclerc) [inline driver/main_args.ml]: cf. @diml [comment](https://github.com/ocaml/ocaml/pull/1945/files#r211238698)

**Comment by gasche** [comment]
I fixed the two glitches that @diml and @xclerc reported. Thanks!

Note that I also need to edit the manpages and the manual with the new option. But before doing this I would like to get an approval in principle first, a sense of consensus (or at least the existence of the opinion) that the feature is desirable.

**Comment by ghost** [comment]
Personally, this option seems good to me. I'm in favor of adding it.

**Comment by gasche** [comment]
Thanks! I updated the PR to include manpage and manual documentation.

File: man/ocamlc.m
Comment (xclerc) [inline man/ocamlc.m]: This change looks suspicious.

Comment (gasche) [reply man/ocamlc.m]: Indeed... Fixed, thanks. (I had tried to get rid of single apostrophes in the text that Emacs groff-mode has trouble with, but I'm not sure what's the proper way to escape them so I gave up in between.)

File: driver/main.ml
Comment (ghost) [inline driver/main.ml]: I suggest to use something like `String.concat "|" P.names` rather than hard-code the list in various places. This way things are more likely to stay in sync when we add more constructors to `P.t`.

File: driver/main_args.ml
Comment (ghost) [inline driver/main_args.ml]: We could also use `Arg.Symbol (Compile_pass.names, f)`, this way we don't need to validate the input in `f`

**Review by ghost — CHANGES_REQUESTED** [review]

File: driver/main.ml
Comment (gasche) [reply driver/main.ml]: We still have those present in the documentation, which is not procedurally generated. But your idea works well in combination with `Arg.Symbol`, will do, thanks.

**Comment by gasche** [comment]
@diml I changed the code as you suggested. A buildup commit factorizing the behavior on recoverable parsing errors in driver/compenv.ml snuck in as 4370d84c9dd5209e6cdeb64d5ae6be7f94c2ea1e.

File: driver/main.ml
Comment (ghost) [inline driver/main.ml]: Note that this code can never be reached now that we use `Symbol`, so we could effectively just use `assert false`

File: driver/main_args.ml
Comment (ghost) [inline driver/main_args.ml]: We can remove the listing in the doc string, the `Arg` module already generates it

File: driver/optmain.ml
Comment (ghost) [inline driver/optmain.ml]: This is dead code as well

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
> A buildup commit factorizing the behavior on recoverable parsing errors in driver/compenv.ml snuck in as 4370d84

ack

**Comment by gasche** [comment]
@diml I made the changes in a new commit (I will rebase later), but I'm having second thoughts about the use of `Arg.Symbol` here. My worry is that people may use `-stop-after ...` in their build script, and get a broken build on older OCaml versions that don't support the pass they use. With my previous implementation, we actually had a soft warning and the option was ignored in that case, which I believe is a safe failover.

**Comment by gasche** [comment]
@diml: I would be interested by getting this approved if you have time for a last bit of ping-pong, as this is a dependency of #292. Thanks a lot for the copious amount of feedback you have given so far.

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
No problem. I see your point about the soft warning but personally I think a hard error is better: it is a bit more work if you want to support older versions of the compiler, but at least there are no surprises. For instance a script using `-stop-after parsing` might make the assumption that the compiler creates no file and execute it in a place where it is not allowed to create files.

**Pull request merged**
